### PR TITLE
feat: add ProductCarousel component

### PR DESCRIPTION
**Pull request by chanceaclark**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,7 +38,7 @@
     "focus-trap-react": "^10.2.3",
     "lodash.debounce": "^4.0.8",
     "lucide-react": "^0.341.0",
-    "next": "14.0.1",
+    "next": "^14.1.4",
```
Comment (joshuawootonn) [inline package.json:41]: I was getting this error when I used latest. https://makeswifthq.slack.com/archives/C014SCS4G75/p1711048297151889

Please make sure you can open the preview URL in the builder before you ship this. Hopefully it works.

File: components/ui/product-preview/product-preview.tsx [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import Image from 'next/image';
-import Link from 'next/link';
 import { forwardRef, Ref } from 'react';
 
+import { Link } from '~/components/link';
```
Comment (joshuawootonn) [inline components/ui/product-preview/product-preview.tsx:4]: Nice. Wasn't sure if I could just swap those. Glad you fixed this.

File: components/contentful-blog-post-list/contentful-blog-post-list.makeswift.tsx [hunk 3]
```diff
@@ -14,7 +14,7 @@ export const props = {
     label: 'Tag',
     async getOptions(query) {
       const response = await fetch('/api/contentful/tags');
-      const body = await response.json();
+      const body: unknown = await response.json();
```
Comment (joshuawootonn) [inline components/contentful-blog-post-list/contentful-blog-post-list.makeswift.tsx:17]: Was there a eslint warning for this?

File: app/api/categories/route.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+const flattenCategoryTree = (categories: CategoryWithChildren[]) => {
+  return categories.reduce<Category[]>((acc, category) => {
+    acc.push(category);
+
+    if (category.children?.length) {
+      acc.push(...flattenCategoryTree(category.children));
+    }
+
+    return acc;
+  }, []);
```
Comment (joshuawootonn) [inline app/api/categories/route.ts:22]: I feel like every time I have tried adding a type to reduce I have had to cast my empty initial array. 🤔 

```diff
- }, []);
+ }, [] as Category[]);
```

Glad you didn't have to.

File: components/product-carousel/product-carousel.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+type QuickSearchResults = Awaited<ReturnType<typeof getQuickSearchResults>>;
+
+export const fetchQuickSearch = async ({ categoryId }: { categoryId?: number }) => {
+  const response = await fetch(
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    `/api/quick-search?categoryId=${categoryId!}&imageHeight=500&imageWidth=500&limit=10`,
+  );
+
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  const { products } = (await response.json()) as QuickSearchResults;
```
Comment (joshuawootonn) [inline components/product-carousel/product-carousel.tsx:36]: Is it feasible to create a zod schema here and parse this?

File: components/product-carousel/product-carousel.tsx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+interface Props {
+  title: string;
+  categoryId?: number;
+}
+
+type QuickSearchResults = Awaited<ReturnType<typeof getQuickSearchResults>>;
+
+export const fetchQuickSearch = async ({ categoryId }: { categoryId?: number }) => {
+  const response = await fetch(
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```
Comment (joshuawootonn) [inline components/product-carousel/product-carousel.tsx]: Maybe you could early return if this is undefined?

File: components/contentful-blog-post-list/contentful-blog-post-list.makeswift.tsx (hunk 3 above)
Comment (chanceaclark) [reply components/contentful-blog-post-list/contentful-blog-post-list.makeswift.tsx:17]: `.json()` returns `any`, so if you type body as `unknown` it should work as is. There is also a package that does this implicitly for us: https://www.npmjs.com/package/@total-typescript/ts-reset

File: app/api/categories/route.ts (hunk 4 above)
Comment (chanceaclark) [reply app/api/categories/route.ts:22]: I use the `.reduce<Category[]>` generic to type it instead

File: components/product-carousel/product-carousel.tsx (hunk 5 above)
Comment (chanceaclark) [reply components/product-carousel/product-carousel.tsx:36]: Yeah... but I was a little lazy here and just followed what was going on in this file: https://github.com/makeswift/catalyst-demo/blob/main/components/cart/client.tsx#L15

File: components/product-carousel/product-carousel.tsx (hunk 6 above)
Comment (chanceaclark) [reply components/product-carousel/product-carousel.tsx]: That completely slipped my mind 🤦‍♂️

File: package.json (hunk 1 above)
Comment (chanceaclark) [reply package.json:41]: Looks good to me

**Pull request merged**